README.md
<p align="center">
  <span>English</span> |
  <a href="README.ru.md#apisite">Pусский</a>
</p>

---

# apisite
> Platform for websites with API which accessible from templates and external clients

[![GoCard][gc1]][gc2]
 [![GitHub Release][gr1]][gr2]
 [![GitHub code size in bytes][sz]]()
 [![Docker Image][di1]][di2]
 [![GitHub license][gl1]][gl2]

[gc1]: https://goreportcard.com/badge/apisite/apisite
[gc2]: https://goreportcard.com/report/github.com/apisite/apisite
[gr1]: https://img.shields.io/github/release/apisite/apisite.svg
[gr2]: https://github.com/apisite/apisite/releases
[sz]: https://img.shields.io/github/languages/code-size/apisite/apisite.svg
[di1]: https://img.shields.io/badge/docker-apisite/apisite-blue.svg
[di2]: https://hub.docker.com/r/apisite/apisite/
[gl1]: https://img.shields.io/github/license/apisite/apisite.svg
[gl2]: LICENSE

## License

The MIT License (MIT), see [LICENSE](LICENSE).

Copyright (c) 2018 Aleksei Kovrizhkin <lekovr+apisite@gmail.com>



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.6.12] - 2023-12-20

### Update deps

* bump versions: golang.org/x/{crypto,net} v0.17.0
* bump versions: go 1.21, gin 1.9.1
* add changelog target
* upd docker workflow deps, add build for arm64v8
* Create codeql-analysis.yml

## [0.2] - 2018-09-09

### Added

* Proof-of-Concept code is working


meta_test.go
package main

import (
	"github.com/pkg/errors"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"net/http"
	"testing"
)

func TestNewMeta(t *testing.T) {

	m := NewMeta(http.StatusNotImplemented, "text/plain")
	assert.Equal(t, http.StatusNotImplemented, m.Status())
	assert.Equal(t, "", m.ErrorMessage(), "no error")

	msg := "Interrupted"
	_, err := m.Raise(http.StatusNotAcceptable, true, msg)
	assert.Equal(t, msg, err.Error(), "err message")
	assert.Equal(t, msg, m.ErrorMessage(), "error message")
	assert.Equal(t, http.StatusNotAcceptable, m.Status(), "status")
}

func TestRaiseNoAbort(t *testing.T) {

	m := Meta{}

	msg := "Not Interrupted"
	_, err := m.Raise(http.StatusNotAcceptable, false, msg)
	require.NoError(t, err, "store error without ebort")
	assert.Equal(t, msg, m.ErrorMessage(), "error message")
	assert.Equal(t, http.StatusNotAcceptable, m.Status(), "status")
}

func TestErrorMessageExternal(t *testing.T) {
	m := Meta{}
	msg := "ExternalError"
	e := errors.New(msg)
	m.SetError(e)
	assert.Equal(t, msg, m.ErrorMessage())
}

func TestRedirectFound(t *testing.T) {
	m := Meta{}
	_, err := m.RedirectFound("/redirect")

	assert.Equal(t, "Abort with redirect", err.Error())
	assert.Equal(t, "/redirect", m.Location())
}


router_test.go
package main

import (
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/stretchr/testify/assert"
	//"github.com/stretchr/testify/require"
)

func TestRouter(t *testing.T) {
	cfg := &Config{}
	log := setupLog()
	r, err := setupRouter(cfg, log)
	assert.NotNil(t, err)
	//	require.NoError(t, err)
	req, _ := http.NewRequest("GET", "/", nil)
	resp := httptest.NewRecorder()
	if err == nil {
		r.ServeHTTP(resp, req)

	}
	assert.Equal(t, resp.Body.String(), "")
}


app/app_test.go
package app

import (
	"net/http"
	"net/http/httptest"
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
	//"github.com/stretchr/testify/require"
)

func TestRun(t *testing.T) {
	// Save original args
	a := os.Args

	tests := []struct {
		name string
		code int
		args []string
	}{
		{"Help", 3, []string{"-h"}},
		{"UnknownFlag", 2, []string{"--unknown"}},
		{"UnknownPort", 1, []string{"--http_addr", ":xx"}},
	}
	for _, tt := range tests {
		os.Args = append([]string{a[0]}, tt.args...)
		var c int
		Run(func(code int) { c = code })
		assert.Equal(t, tt.code, c, tt.name)
	}

	// Restore original args
	os.Args = a
}

func TestRouter(t *testing.T) {
	cfg := &Config{}
	log := setupLog()
	r, err := setupRouter(cfg, log)
	assert.NotNil(t, err)
	//	require.NoError(t, err)
	req, _ := http.NewRequest("GET", "/", nil)
	resp := httptest.NewRecorder()
	if err == nil {
		r.ServeHTTP(resp, req)

	}
	assert.Equal(t, resp.Body.String(), "")
}


main.go
// +build !test

// This file holds code which does not covered by tests

/*
Package main contains only `app.Run(os.Exit)` call.

See https://godoc.org/github.com/apisite/apisite/app for details.

Command options:

  $ ./apisite -h
  apisite v0.6.1-6-ga26c508. API website engine

Usage:
  apisite [OPTIONS]

Application Options:
      --http_addr=      Http listen address (default: localhost:8080)
      --content_type=   Default content type (default: text/html; charset=utf-8)
      --error_404=      Template called when page is not found (default: .404)
      --buffer_size=    Template buffer size (default: 64)

Filesystem Options:
      --fs.templates=   Templates root path (default: tmpl/)
      --fs.mask=        Templates filename mask (default: .tmpl)
      --fs.includes=    Includes path (default: inc/)
      --fs.layouts=     Layouts path (default: layout/)
      --fs.pages=       Pages path (default: page/)
      --fs.use_suffix   Template type defined by suffix
      --fs.index=       Index page name (default: index)
      --fs.def_layout=  Default layout template (default: default)
      --fs.hide_prefix= Treat files with this prefix as hidden (default: .)

API Options:
      --api.dsn=        Database connect string (default: postgres://?sslmode=disable)
      --api.driver=     Database driver (default: postgres)
      --api.indef=      Argument definition function (default: func_args)
      --api.outdef=     Result row definition function (default: func_result)
      --api.index=      Available functions list (default: index)
      --api.schema=     Definition functions schema (default: rpc)
      --api.arg_syntax= Default named args syntax (:= or =>) (default: :=)
      --api.arg_prefix= Trim prefix from arg name (default: a_)
      --api.nsp=        Proc namespace(s)

Help Options:
  -h, --help            Show this help message

*/
package main

import (
	"github.com/apisite/apisite/app"
	"log"
	"os"
)

// Actual version value will be set at build time
var version = "0.0-dev"

func main() {
	app.Version = version
	log.Printf("apisite %s. API website engine", version)
	app.Run(os.Exit)
}


meta.go
package main

import (
	"github.com/gin-gonic/gin"
	"html/template"

	base "github.com/apisite/apitpl/ginapitpl/samplemeta"
)

// Meta holds template metadata
type Meta struct {
	base.Meta
	JS           []string
	CSS          []string
}

// NewMeta returns new initialised Meta struct
func NewMeta(status int, ctype string) *Meta {
	m := base.NewMeta(status, ctype)
	return &Meta{Meta: *m}
}

// AddJS - add .js file to scripts list
func (p *Meta) AddJS(file string) (string, error) {
	p.JS = append(p.JS, file)
	return "", nil
}

// AddCSS - add .css file to styles list
func (p *Meta) AddCSS(file string) (string, error) {
	p.JS = append(p.CSS, file)
	return "", nil
}

// SetProtoFuncs appends function templates and not related to request functions to funcs
func SetProtoFuncs(funcs template.FuncMap) {
	funcs["request"] = func() interface{} { return nil }
	funcs["param"] = func(key string) string { return "" }
}

// SetRequestFuncs appends funcs which return real data inside request processing
func SetRequestFuncs(funcs template.FuncMap, ctx *gin.Context) {
	funcs["request"] = func() interface{} { return ctx.Request }
	funcs["param"] = func(key string) string { return ctx.Param(key) }
}


pager.go
package main

import (
	"strconv"
)

type Pager struct {
	Count   int
	Max     int
	Current int
	Enabled bool
	IsFirst bool
	IsLast  bool
	Prev    string
	Next    string
}

func pagerInit(itemCount, pageSize, pageNum int, prefix string) Pager {
	pages := 0
	if pageSize > 0 {
		pages = int(itemCount / pageSize)
		if itemCount%pageSize > 0 {
			pages++
		}
	}
	max := pages - 1
	prev := "#"
	if pageNum == 1 {
		prev = "../" + prefix + "/"
	} else if pageNum > 1 {
		prev = strconv.Itoa(pageNum-1) + ".html"
	}
	next := "#"
	if pageNum < max {
		next = strconv.Itoa(pageNum+1) + ".html"
	}

	// ------------------------
	p := Pager{
		Count:   pages,
		Max:     max,
		Current: pageNum,
		Enabled: (pages > 0),
		IsFirst: (pageNum == 0),
		IsLast:  (pageNum == max),
		Prev:    prev,
		Next:    next,
	}
	return p
}


router.go
package main

import (
	"context"
	"github.com/pkg/errors"
	"html/template"
	"net/http"
	"os"
	"os/signal"
	"time"

	mapper "github.com/birkirb/loggers-mapper-logrus"
	//	"github.com/onrik/logrus/filename"
	"github.com/sirupsen/logrus"
	"gopkg.in/birkirb/loggers.v1"

	"github.com/jessevdk/go-flags"

	"github.com/gin-contrib/static"
	"github.com/gin-gonic/gin"

	"github.com/apisite/procapi"
	"github.com/apisite/procapi/ginproc"

	"github.com/apisite/procapi/pgtype"
	_ "github.com/lib/pq"

	"github.com/apisite/apitpl"
	"github.com/apisite/apitpl/ginapitpl"
	"github.com/apisite/apitpl/lookupfs"
)

// Config holds all config vars
type Config struct {
	Addr        string `long:"http_addr" default:"localhost:8080"  description:"Http listen address"`
	ContentType string `long:"content_type" default:"text/html; charset=utf-8" description:"Default content type"`
	Error404    string `long:"error_404" default:".404" description:"Template called when page is not found"`
	BufferSize  int    `long:"buffer_size" default:"64" description:"Template buffer size"`

	FS  lookupfs.Config `group:"Filesystem Options" namespace:"fs" env-namespace:"FS"`
	API procapi.Config  `group:"API Options" namespace:"api" env-namespace:"API"`
}

func setupConfig() (*Config, error) {
	cfg := &Config{}
	p := flags.NewParser(cfg, flags.Default)
	if _, err := p.Parse(); err != nil {
		if e, ok := err.(*flags.Error); ok && e.Type == flags.ErrHelp {
			return nil, errors.New("ERR1") //os.Exit(1) // help printed
		} else {
			return nil, errors.New("ERR2") //os.Exit(2) // error message written already
		}
	}
	return cfg, nil
}

func setupLog() loggers.Contextual {
	l := logrus.New()

	if gin.IsDebugging() {
		l.SetLevel(logrus.DebugLevel)
		l.SetReportCaller(true)
		//l.AddHook(filename.NewHook())
	}
	return mapper.NewLogger(l)
}

func setupRouter(cfg *Config, log loggers.Contextual) (*gin.Engine, error) {

	// procapi

	db := procapi.New(cfg.API, log, nil).SetMarshaller(pgtype.New())
	err := db.Open()
	if err != nil {
		return nil, err
	}
	err = db.LoadMethods()
	if err != nil {
		return nil, err
	}
	api := ginproc.NewServer(log, db)

	// apitpl

	allFuncs := make(template.FuncMap, 0)
	SetSimpleFuncs(allFuncs)
	SetProtoFuncs(allFuncs)
	api.SetProtoFuncs(allFuncs)

	//	gin.SetMode(gin.ReleaseMode)

	r := gin.Default()

	fs := lookupfs.New(cfg.FS)
	tfs, err := apitpl.New(cfg.BufferSize).Funcs(allFuncs).LookupFS(fs).Parse()
	if err != nil {
		return nil, err
	}
	tfs.ParseAlways(gin.IsDebugging())
	gintpl := ginapitpl.New(log, tfs)
	gintpl.RequestHandler = func(ctx *gin.Context, funcs template.FuncMap) ginapitpl.MetaData {
		SetRequestFuncs(funcs, ctx)
		api.SetRequestFuncs(funcs, ctx)
		m := NewMeta(http.StatusOK, cfg.ContentType)
		m.SetLayout(cfg.FS.DefLayout)
		return m
	}

	gintpl.Route("", r)

	r.Use(static.Serve("/", static.LocalFile("./static", false)))
	r.NoRoute(func(c *gin.Context) {
		gintpl.HTML(c, cfg.Error404)
	})

	api.Route("/rpc", r)

	return r, nil
}

func runServer(log loggers.Contextual, addr string, r *gin.Engine) error {

	srv := &http.Server{
		Addr:    addr,
		Handler: r,
	}

	go func() {
		// service connections
		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			log.Fatal(err)
		}
	}()

	// Wait for interrupt signal to gracefully shutdown the server with
	// a timeout of 5 seconds.
	quit := make(chan os.Signal)
	signal.Notify(quit, os.Interrupt)
	<-quit
	log.Println("Shutdown Server ...")

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	if err := srv.Shutdown(ctx); err != nil {
		log.Fatal("Server Shutdown:", err)
	}
	log.Println("Server stopped")
	return nil
}


tplfunc/funcmap.go
package tplfunc

import (
	//	"fmt"
	"html/template"
	"net/url"
	"strconv"

	"github.com/acoshift/paginate"
	"github.com/spf13/cast"
)

// Page holds pagination functionality
type Page struct {
	ID      int64 // Page number
	IsPrev  bool  // Is it a 'prev' link
	IsNext  bool  // Is it a 'last' link
	Allowed bool
	Href    string // Page href

	//	IsCurrent   bool // Is this a current page
	//	IsFirst     bool // Is it a first page
	//	IsLast      bool // Is it a last page
}

func valWithDefaultInt64(args url.Values, key string, valDefault int64) int64 {
	a, ok := args[key]
	if !ok {
		return valDefault
	}
	if len(a) == 0 {
		return valDefault
	}
	return cast.ToInt64(a[0])
}

// Pager returns array with pagination links
func Pager(args url.Values, count interface{}, argPrefix string, rowsMax, around, edge int64) *[]Page {
	itemCount := cast.ToInt64(count)

	rowLimit := valWithDefaultInt64(args, argPrefix+"lim", rowsMax)
	rowOffset := valWithDefaultInt64(args, argPrefix+"off", 0)

	pn := paginate.FromLimitOffset(rowLimit, rowOffset, itemCount)
	pageCurrent := rowOffset/rowLimit + 1
	p := []Page{}
	p = append(p, Page{IsPrev: true, Allowed: pn.CanPrev(), ID: pn.Prev(), Href: ""})
	for _, pi := range pn.Pages(around, edge) {
		p = append(p, Page{Allowed: pi != pageCurrent, ID: pi, Href: ""})
	}
	p = append(p, Page{IsNext: true, Allowed: pn.CanNext(), ID: pn.Next(), Href: ""})

	if rowLimit != rowsMax {
		args.Set(argPrefix+"lim", cast.ToString(rowLimit))
	} else {
		args.Del(argPrefix + "lim")
	}

	for i := range p {
		if p[i].ID > 1 {
			args.Set(argPrefix+"off", cast.ToString(rowLimit*(p[i].ID-1)))
		} else {
			args.Del(argPrefix + "off")
		}
		p[i].Href = "?" + args.Encode()
	}
	return &p
}

func interator(min, max int) []int {
	var ret []int
	if max == 0 {
		return ret
	}
	for i := min; i <= max; i++ {
		ret = append(ret, i)
	}
	return ret
}

func add(a, b int) int {
	return a + b
}

func atoi(a string, def int) int {
	if a == "" {
		return def
	}
	rv, err := strconv.Atoi(a)
	if err != nil {
		return def
	}
	return rv

}

// SetSimpleFuncs registers all prevoious funcs in given FuncMap
// This func is not for use in templates
func SetSimpleFuncs(funcs template.FuncMap) {
	funcs["add"] = add
	funcs["atoi"] = atoi
	funcs["pager"] = Pager
	funcs["interator"] = interator
	funcs["bool"] = func(a bool) string {
		if a {
			return "+"
		}
		return "-"
	}
	funcs["HTML"] = func(s string) template.HTML {
		return template.HTML(s)
	}
	funcs["ref"] = func(a *interface{}) interface{} {
		return *a
	}
}


tplfunc/tplfunc.go
// Package tplfunc implements functions avalilable inside templates
package tplfunc

import (
	"github.com/gin-gonic/gin"
	"html/template"

	base "github.com/apisite/apitpl/ginapitpl/samplemeta"
)

// Meta holds template metadata
// This func is not for use in templates
type Meta struct {
	base.Meta
	JS  []string
	CSS []string
}

// NewMeta returns new initialised Meta struct
// This func is not for use in templates
func NewMeta(status int, ctype string) *Meta {
	m := base.NewMeta(status, ctype)
	return &Meta{Meta: *m}
}

// AddJS - add .js file to scripts list
func (p *Meta) AddJS(file string) (string, error) {
	p.JS = append(p.JS, file)
	return "", nil
}

// AddCSS - add .css file to styles list
func (p *Meta) AddCSS(file string) (string, error) {
	p.JS = append(p.CSS, file)
	return "", nil
}

// SetProtoFuncs appends function templates and not related to request functions to funcs
// This func is not for use in templates
func SetProtoFuncs(funcs template.FuncMap) {
	funcs["request"] = func() interface{} { return nil }
	funcs["param"] = func(key string) string { return "" }
}

// SetRequestFuncs appends funcs which return real data inside request processing
// This func is not for use in templates
func SetRequestFuncs(funcs template.FuncMap, ctx *gin.Context) {
	funcs["request"] = func() interface{} { return ctx.Request }
	funcs["param"] = func(key string) string { return ctx.Param(key) }
}


app/app.go
// Package app implements apisite command functions
package app

import (
	"context"
	"github.com/pkg/errors"
	"html/template"
	"log"
	"net/http"
	"os"
	"os/signal"
	"time"

	mapper "github.com/birkirb/loggers-mapper-logrus"
	//	"github.com/onrik/logrus/filename"
	"github.com/sirupsen/logrus"
	"gopkg.in/birkirb/loggers.v1"

	"github.com/jessevdk/go-flags"

	"github.com/gin-contrib/static"
	"github.com/gin-gonic/gin"

	"github.com/apisite/procapi"
	"github.com/apisite/procapi/ginproc"

	"github.com/apisite/apisite/tplfunc"
	"github.com/apisite/apitpl"
	"github.com/apisite/apitpl/ginapitpl"
	"github.com/apisite/apitpl/lookupfs"
)

// Config holds all config vars
type Config struct {
	Addr        string `long:"http_addr" default:"localhost:8080"  description:"Http listen address"`
	ContentType string `long:"content_type" default:"text/html; charset=utf-8" description:"Default content type"`
	Error404    string `long:"error_404" default:".404" description:"Template called when page is not found"`
	BufferSize  int    `long:"buffer_size" default:"64" description:"Template buffer size"`

	FS  lookupfs.Config `group:"Filesystem Options" namespace:"fs" env-namespace:"FS"`
	API procapi.Config  `group:"API Options" namespace:"api" env-namespace:"API"`
}

var (
	// ErrGotHelp returned after showing requested help
	ErrGotHelp = errors.New("help printed")
	// ErrBadArgs returned after showing command args error message
	ErrBadArgs = errors.New("option error printed")
)

// Version value will be set in main()
var Version = "0.0-dev"

// Run called by main() and prerforms all of app
func Run(exitFunc func(code int)) {
	var err error
	var cfg *Config
	defer func() { shutdown(exitFunc, err) }()
	cfg, err = setupConfig()
	if err != nil {
		return
	}
	l := setupLog()
	var r *gin.Engine
	r, err = setupRouter(cfg, l)
	if err != nil {
		return
	}
	err = runServer(l, cfg.Addr, r)
}

// exit after deferred cleanups have run
func shutdown(exitFunc func(code int), e error) {
	if e != nil {
		var code int
		switch e {
		case ErrGotHelp:
			code = 3
		case ErrBadArgs:
			code = 2
		default:
			code = 1
			log.Printf("Run error: %s", e.Error())
		}
		exitFunc(code)
	}
}

func setupConfig() (*Config, error) {
	cfg := &Config{}
	p := flags.NewParser(cfg, flags.Default)
	if _, err := p.Parse(); err != nil {
		if e, ok := err.(*flags.Error); ok && e.Type == flags.ErrHelp {
			return nil, ErrGotHelp
		}
		return nil, ErrBadArgs
	}
	return cfg, nil
}

// setupLog creates logger
func setupLog() loggers.Contextual {
	l := logrus.New()
	if gin.IsDebugging() {
		l.SetLevel(logrus.DebugLevel)
		l.SetReportCaller(true)
	}
	return &mapper.Logger{Logger: l} // Same as mapper.NewLogger(l) but without info log message
}

func setupRouter(cfg *Config, log loggers.Contextual) (*gin.Engine, error) {

	// procapi

	db := procapi.New(cfg.API, log, nil)
	err := db.Open()
	if err != nil {
		return nil, err
	}
	err = db.LoadMethods()
	if err != nil {
		return nil, err
	}
	api := ginproc.NewServer(log, db)

	// apitpl

	allFuncs := template.FuncMap{
		"version": func() string { return Version },
	}
	tplfunc.SetSimpleFuncs(allFuncs)
	tplfunc.SetProtoFuncs(allFuncs)
	api.SetProtoFuncs(allFuncs)

	//	gin.SetMode(gin.ReleaseMode)

	r := gin.Default()

	fs := lookupfs.New(cfg.FS)
	tfs, err := apitpl.New(cfg.BufferSize).Funcs(allFuncs).LookupFS(fs).Parse()
	if err != nil {
		return nil, err
	}
	tfs.ParseAlways(gin.IsDebugging())
	gintpl := ginapitpl.New(log, tfs)
	gintpl.RequestHandler = func(ctx *gin.Context, funcs template.FuncMap) ginapitpl.MetaData {
		tplfunc.SetRequestFuncs(funcs, ctx)
		api.SetRequestFuncs(funcs, ctx)
		m := tplfunc.NewMeta(http.StatusOK, cfg.ContentType)
		m.SetLayout(cfg.FS.DefLayout)
		return m
	}

	gintpl.Route("", r)

	r.Use(static.Serve("/", static.LocalFile("./static", false)))
	r.NoRoute(func(c *gin.Context) {
		gintpl.HTML(c, cfg.Error404)
	})

	api.Route("/rpc", r) // nolint: errcheck

	return r, nil
}

func runServer(log loggers.Contextual, addr string, r *gin.Engine) error {

	srv := &http.Server{
		Addr:    addr,
		Handler: r,
	}

	go func() {
		// service connections
		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			log.Fatal(err)
		}
	}()

	// Wait for interrupt signal to gracefully shutdown the server with
	// a timeout of 5 seconds.
	quit := make(chan os.Signal, 1)
	signal.Notify(quit, os.Interrupt)
	<-quit
	log.Println("Shutdown Server ...")

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	if err := srv.Shutdown(ctx); err != nil {
		log.Fatal("Server Shutdown:", err)
	}
	log.Println("Server stopped")
	return nil
}
